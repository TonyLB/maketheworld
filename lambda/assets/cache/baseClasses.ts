import { ComponentRenderItem } from "@tonylb/mtw-wml/dist/normalize"

export type EphemeraCondition = {
    dependencies: string[];
    if: string;
}

export type EphemeraFeatureAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
}

export type EphemeraFeature = {
    EphemeraId: string;
    key: string;
    tag: 'Feature';
    appearances: EphemeraFeatureAppearance[];
}

export type EphemeraExit = {
    name: string;
    to: string;
}

export type EphemeraRoomAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
    exits: EphemeraExit[];
}

export type EphemeraRoom = {
    EphemeraId: string;
    key: string;
    tag: 'Room';
    appearances: EphemeraRoomAppearance[];
}

export type EphemeraMapRoom = {
    EphemeraId: string;
    x: number;
    y: number;
}

export type EphemeraMapAppearance = {
    conditions: EphemeraCondition[];
    fileURL: string;
    name: string;
    rooms: Record<string, EphemeraMapRoom>;
}

export type EphemeraMap = {
    EphemeraId: string;
    key: string;
    tag: 'Map';
    appearance: EphemeraMapAppearance;
}

//
// TODO (ISS1385): Add EphemeraAction to base classes
//

export type EphemeraItem = EphemeraFeature | EphemeraRoom | EphemeraMap
