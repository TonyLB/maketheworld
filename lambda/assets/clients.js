import { S3Client, GetObjectCommand as GetObjectCommandOriginal } from "@aws-sdk/client-s3"
import { EventBridgeClient } from "@aws-sdk/client-eventbridge"

export const ebClient = new EventBridgeClient({ region: process.env.AWS_REGION })

// export const s3Client = new S3Client(params)

export class GetObjectCommand extends GetObjectCommandOriginal {}
