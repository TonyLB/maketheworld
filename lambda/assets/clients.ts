import { GetObjectCommand as GetObjectCommandOriginal } from "@aws-sdk/client-s3"
import { EventBridgeClient, PutEventsCommand } from "@aws-sdk/client-eventbridge"
import { apiClient as apiClientImport } from "@tonylb/mtw-utilities/dist/apiManagement/apiManagementClient"
import { SFNClient } from "@aws-sdk/client-sfn"

export const ebClient = new EventBridgeClient({ region: process.env.AWS_REGION })
export const sfnClient = new SFNClient({ region: process.env.AWS_REGION })

// export const s3Client = new S3Client(params)

export class GetObjectCommand extends GetObjectCommandOriginal {}

export const apiClient = {
    send: (message: any) => {
        return apiClientImport.send(message)
            .catch(async (err: any) => {
                if (err.name === 'GoneException' || err.name === 'BadRequestException') {
                    await ebClient.send(new PutEventsCommand({
                        Entries: [{
                            EventBusName: process.env.EVENT_BUS_NAME,
                            Source: 'mtw.diagnostics',
                            DetailType: 'Force Disconnect',
                            Detail: JSON.stringify({
                                connectionId: message.ConnectionId
                            })
                        }]
                    }))
                }
                else {
                    console.log(`Error: ${err.name}`)
                    throw err
                }
            })
    }
}
