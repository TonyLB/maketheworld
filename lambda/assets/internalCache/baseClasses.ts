export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>
