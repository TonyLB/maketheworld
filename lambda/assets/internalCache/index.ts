import { connectionDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index'
import { delayPromise } from '@tonylb/mtw-utilities/dist/dynamoDB/delayPromise'

export class CacheBase {
    async get(props: { category: never; key: never }) {
        return undefined
    }
    async clear() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

type CacheConstructor = Constructor<{
    get(props: { category: never; key: never; }): Promise<undefined>;
    clear(): void;
}>

type CacheGlobalData = {
    connectionId: string;
    RequestId: string;
}

export const CacheGlobal = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheGlobal extends Base {
        Global: Partial<CacheGlobalData> = {}

        override get(props: { category: 'Global'; key: keyof CacheGlobalData}): Promise<CacheGlobalData[typeof props["key"]]>
        override get(...args: Parameters<InstanceType<GBase>["get"]>): ReturnType<InstanceType<GBase>["get"]>
        override async get(props: any) {
            if (props.category === 'Global') {
                return this.Global[props.key]
            }
            return await super.get(props)
        }

        override clear() {
            this.Global = {}
            super.clear()
        }

        set(props: { category: 'Global', key: 'connectionId' | 'RequestId', value: string }): void
        set({ key, value }: { category: 'Global', key: keyof CacheGlobalData, value: any }): void {
            this.Global[key] = value
        }
    }
}

type CacheLookupOnceData = {
    player: string;
}

export const CacheLookupOnce = <GBase extends Constructor<{ get(props: { category: 'Global', key: 'connectionId' }): Promise<string>; clear(): void; }>>(Base: GBase) => {
    return class CacheGlobal extends Base {
        Lookup: CacheLookupOnceData | undefined = undefined

        override get(props: { category: 'Lookup'; key: 'player'}): Promise<string>
        override get(...args: Parameters<InstanceType<GBase>["get"]>): ReturnType<InstanceType<GBase>["get"]>
        override async get(props: any) {
            if (props.category === 'Lookup') {
                if (this.Lookup === undefined) {
                    const connectionId = await super.get({
                        category: 'Global',
                        key: 'connectionId'
                    })
                    if (connectionId) {
                        //
                        // TODO: Replace repeated attempts with exponential backoff by
                        // refactoring connectionDB.getItem to allow a consistent argument
                        // that can actviate strongly-consistent reads
                        //
                        let attempts = 0
                        let exponentialBackoff = 50
                        while(attempts < 5) {
                            const { player = '' } = await connectionDB.getItem<{ player: string }>({
                                ConnectionId: connectionId,
                                DataCategory: 'Meta::Connection',
                                ProjectionFields: ['player']
                            }) || {}
                            if (player) {
                                return { player }
                            }
                            attempts += 1
                            await delayPromise(exponentialBackoff)
                            exponentialBackoff = exponentialBackoff * 2
                        }
                    }
                }
                return this.Lookup?.[props.key]
            }
            return await super.get(props)
        }

        override clear() {
            this.Lookup = undefined
            super.clear()
        }

    }
}

const InternalCache = CacheLookupOnce(CacheGlobal(CacheBase))
export const internalCache = new InternalCache()
export default internalCache
