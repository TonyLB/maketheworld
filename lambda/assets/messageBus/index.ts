import {
    MessageBus,
    isFetchLibraryAPIMessage,
    isFetchAssetAPIMessage,
    isUploadResponseMessage
} from "./baseClasses"
import fetchLibraryMessage from "../fetchLibrary"
import fetchAssetMessage from "../fetch"
import { uploadResponseMessage } from "../upload/uploadResponse"

export const messageBus = new MessageBus()

messageBus.subscribe({
    tag: 'FetchLibrary',
    priority: 5,
    filter: isFetchLibraryAPIMessage,
    callback: fetchLibraryMessage
})
messageBus.subscribe({
    tag: 'FetchAsset',
    priority: 5,
    filter: isFetchAssetAPIMessage,
    callback: fetchAssetMessage
})
messageBus.subscribe({
    tag: 'UploadResponse',
    priority: 8,
    filter: isUploadResponseMessage,
    callback: uploadResponseMessage
})

export default messageBus