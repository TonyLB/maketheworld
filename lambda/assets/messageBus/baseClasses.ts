import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'
import { AssetWorkspaceAddress } from '@tonylb/mtw-asset-workspace/dist'
import { AssetPlayerSettingsAPIMessage, ParseWMLAPIImage } from '@tonylb/mtw-interfaces/dist/asset';
import { EphemeraAssetId, EphemeraCharacterId } from '@tonylb/mtw-interfaces/dist/baseClasses';
import { LibraryAsset, LibraryCharacter } from '@tonylb/mtw-interfaces/dist/library';

export type ReturnValueMessage = {
    type: 'ReturnValue';
    body: Record<string, any>;
}

export type FetchLibraryMessage = {
    type: 'FetchLibrary';
}

type ImportFromAssetArgument = {
    assetId: `ASSET#${string}`;
    keys: string[];
}

export type FetchImportsMessage = {
    type: 'FetchImports';
    importsFromAsset: ImportFromAssetArgument[];
}

export type FetchAssetMessage = {
    type: 'FetchAsset';
    AssetId?: string;
    fileName?: string;
}

type UploadURLMessageImage = {
    key: string;
    contentType: string;
}

export type UploadURLMessage = {
    type: 'UploadURL';
    assetType: 'Character' | 'Asset';
    images: UploadURLMessageImage[];
}

//
// TODO: Refactor ParseWMLMessage so that it pulls the AssetId field from the
// upload, rather than requiring a specific AssetWorkspaceAddress.  Add a "create"
// flag to explicitly authorize creation of a new asset file (rather than update),
// and otherwise fail if the Asset doesn't exist.  Fail with access rejection if
// the asset exists but the player is not authorized to update.
//
export type ParseWMLMessage = {
    type: 'ParseWML';
    AssetId: EphemeraCharacterId | EphemeraAssetId;
    uploadName: string;
    images?: ParseWMLAPIImage[];
    create?: boolean;
}

export type FormatImageMessage = {
    type: 'FormatImage';
    fileName: string;
    width: number;
    height: number;
    AssetId: string;
    imageKey: string;
}

export type MoveAssetMessage = {
    type: 'MoveAsset';
    from: AssetWorkspaceAddress;
    to: AssetWorkspaceAddress;
}

export type MoveByAssetIdMessage = {
    type: 'MoveByAssetId',
    AssetId: string;
    toZone: 'Canon' | 'Library' | 'Personal';
    player?: string;
}

export type LibrarySubscribeMessage = {
    type: 'LibrarySubscribe';
}

export type LibraryUnsubscribeMessage = {
    type: 'LibraryUnsubscribe';
}

export type PlayerInfoMessage = {
    type: 'PlayerInfo';
    player?: string;
    RequestId?: string;
}

export type PlayerSettingsMessage = {
    type: 'PlayerSettings';
    player?: string;
    RequestId?: string;
} & Omit<AssetPlayerSettingsAPIMessage, 'message'>

export type LibraryUpdateMessage = {
    type: 'LibraryUpdate';
}

export type RemoveAssetMessage = {
    type: 'RemoveAsset';
    AssetId: string;
}

export type MessageType = ReturnValueMessage |
    FetchLibraryMessage |
    FetchImportsMessage |
    FetchAssetMessage |
    UploadURLMessage |
    FormatImageMessage |
    ParseWMLMessage |
    MoveAssetMessage |
    MoveByAssetIdMessage |
    LibrarySubscribeMessage |
    LibraryUnsubscribeMessage |
    PlayerInfoMessage |
    PlayerSettingsMessage |
    LibraryUpdateMessage |
    RemoveAssetMessage

export const isReturnValueMessage = (prop: MessageType): prop is ReturnValueMessage => (prop.type === 'ReturnValue')
export const isFetchLibraryAPIMessage = (prop: MessageType): prop is FetchLibraryMessage => (prop.type === 'FetchLibrary')
export const isFetchAssetAPIMessage = (prop: MessageType): prop is FetchAssetMessage => (prop.type === 'FetchAsset')
export const isFetchImportsAPIMessage = (prop: MessageType): prop is FetchImportsMessage => (prop.type === 'FetchImports')
export const isUploadURLMessage = (prop: MessageType): prop is UploadURLMessage => (prop.type === 'UploadURL')
export const isFormatImageMessage = (prop: MessageType): prop is FormatImageMessage => (prop.type === 'FormatImage')
export const isParseWMLMessage = (prop: MessageType): prop is ParseWMLMessage => (prop.type === 'ParseWML')
export const isMoveAssetMessage = (prop: MessageType): prop is MoveAssetMessage => (prop.type === 'MoveAsset')
export const isMoveByAssetIdMessage = (prop: MessageType): prop is MoveByAssetIdMessage => (prop.type === 'MoveByAssetId')
export const isLibrarySubscribeMessage = (prop: MessageType): prop is LibrarySubscribeMessage => (prop.type === 'LibrarySubscribe')
export const isLibraryUnsubscribeMessage = (prop: MessageType): prop is LibraryUnsubscribeMessage => (prop.type === 'LibraryUnsubscribe')
export const isPlayerInfoMessage = (prop: MessageType): prop is PlayerInfoMessage => (prop.type === 'PlayerInfo')
export const isPlayerSettingMessage = (prop: MessageType): prop is PlayerSettingsMessage => (prop.type === 'PlayerSettings')
export const isLibraryUpdateMessage = (prop: MessageType): prop is LibraryUpdateMessage => (prop.type === 'LibraryUpdate')
export const isRemoveAssetMessage = (prop: MessageType): prop is RemoveAssetMessage => (prop.type === 'RemoveAsset')

export class MessageBus extends InternalMessageBus<MessageType> {}
