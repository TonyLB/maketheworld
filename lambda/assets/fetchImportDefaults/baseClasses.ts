import { EphemeraAssetId } from "@tonylb/mtw-interfaces/ts/baseClasses"
import { JSONFileCache } from "../internalCache/jsonFile"
import { Graph } from "@tonylb/mtw-utilities/dist/graphStorage/utils/graph"
import { AssetWorkspaceAddress } from "@tonylb/mtw-asset-workspace"
import ReadOnlyAssetWorkspace from "@tonylb/mtw-asset-workspace/dist/readOnly"

export class InheritanceGraph extends Graph<EphemeraAssetId, { key: EphemeraAssetId; address: AssetWorkspaceAddress }, {}> {}

export class FetchImportsJSONHelper {
    _inheritanceGraph: InheritanceGraph
    constructor(inheritanceGraph: InheritanceGraph) {
        this._inheritanceGraph = inheritanceGraph
    }

    async get(assetId: EphemeraAssetId): Promise<JSONFileCache> {
        const node = this._inheritanceGraph.nodes[assetId]
        if (node) {
            const assetWorkspace = new ReadOnlyAssetWorkspace(node.address)
            await assetWorkspace.loadJSON()
            return {
                normal: assetWorkspace.normal || {},
                namespaceIdToDB: assetWorkspace.namespaceIdToDB || {}
            }    
        }
        else {
            return {
                normal: {},
                namespaceIdToDB: {}
            }
        }
    }
}

