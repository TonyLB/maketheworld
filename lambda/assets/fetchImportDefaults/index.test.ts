import Normalizer from '@tonylb/mtw-wml/dist/normalize'
import { fetchImportsMessage } from '.'

jest.mock('../clients')
import { apiClient } from '../clients'
jest.mock('../messageBus')
import messageBus from '../messageBus'
jest.mock('../internalCache')
import internalCache from '../internalCache'
import { NormalForm } from '@tonylb/mtw-wml/dist/normalize/baseClasses'

const messageBusMock = jest.mocked(messageBus, true)
const apiClientMock = apiClient as jest.Mocked<typeof apiClient>
const internalCacheMock = jest.mocked(internalCache, true)

const testNormalFromWML = (wml: string): NormalForm => {
    const normalizer = new Normalizer()
    normalizer.loadWML(wml)
    return normalizer.normal
}

describe('fetchImportsMessage', () => {
    const testFinal = testNormalFromWML(`<Asset key=(testFinal)>
        <Room key=(testNonImport)>
            <Description>
                DescriptionOne
            </Description>
            <Exit to=(testNonImportStub)>test exit</Exit>
        </Room>
        <Room key=(testNonImportStub)>
            <Name>StubOne</Name>
        </Room>
        <Room key=(testImportOne)>
            <Description>
                Two
            </Description>
            <Exit to=(testImportStubOne)>test exit one</Exit>
        </Room>
        <Room key=(testImportStubOne) />
        <Room key=(testImportTwo) />
        <Room key=(testNonImportTwo)>
            <Exit to=(testImportTwo)>test exit</Exit>
        </Room>
        <Room key=(testImportThree) />
        <Import from=(testImportAssetOne)>
            <Use type="Room" key=(testImportOne) />
            <Use type="Room" key=(testImportStubOne) />
            <Use type="Room" key=(testImportFoo) as=(testImportTwo) />
        </Import>
        <Import from=(testImportAssetTwo)>
            <Use type="Room" key=(basic) as=(testImportThree) />
        </Import>
    </Asset>`)
    const testImportOne = testNormalFromWML(`<Asset key=(testImportAssetOne)>
        <Room key=(testImportOne)>
            <Description>
                One
            </Description>
        </Room>
        <Room key=(testImportStubOne)>
            <Name>StubTwo</Name>
        </Room>
        <Room key=(testImportFoo)>
            <Name>StubFoo</Name>
            <Description>
                Foo
            </Description>
        </Room>
    </Asset>`)
    const testImportTwo = testNormalFromWML(`<Asset key=(testImportAssetTwo)>
        <Room key=(basic)>
            <Description>
                Asset Two
            </Description>
            <Exit to=(stub)>test exit</Exit>
        </Room>
        <Room key=(stub) />
        <Import from=(testImportAssetThree)>
            <Use type="Room" key=(basicTwo) as=(stub) />
            <Use type="Room" key=(basicOne) as=(basic) />
        </Import>
    </Asset>`)
    const testImportThree = testNormalFromWML(`<Asset key=(testImportAssetThree)>
        <Room key=(basicOne)>
            <Exit to=(stub)>test exit</Exit>
        </Room>
        <Room key=(basicTwo)><Name>Asset Three</Name></Room>
        <Room key=(stub)><Name>AssetThreeStub</Name></Room>
    </Asset>`)
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        internalCacheMock.JSONFile.get.mockImplementation(async (assetId: string) => {
            let normal: NormalForm = {}
            switch(assetId) {
                case 'ASSET#testFinal':
                    normal = testFinal
                    break
                case 'ASSET#testImportAssetOne':
                    normal = testImportOne
                    break
                case 'ASSET#testImportAssetTwo':
                    normal = testImportTwo
                    break
                case 'ASSET#testImportAssetThree':
                    normal = testImportThree
                    break
            }
            return {
                normal,
                namespaceIdToDB: {}
            }
        })
    })

    it('should return empty when passed no keys', async () => {
        await fetchImportsMessage({ messageBus: messageBusMock, payloads: [{ type: 'FetchImports', importsFromAsset: [{ assetId: 'ASSET#testFinal', keys: [] }] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should return element and stubs when passed non-import key', async () => {
        await fetchImportsMessage({ messageBus: messageBusMock, payloads: [{ type: 'FetchImports', importsFromAsset: [{ assetId: 'ASSET#testFinal', keys: ['testNonImport'] }] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should recursive fetch one level of element and stubs when passed import key', async () => {
        await fetchImportsMessage({ messageBus: messageBusMock, payloads: [{ type: 'FetchImports', importsFromAsset: [{ assetId: 'ASSET#testFinal', keys: ['testImportOne'] }] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should follow dynamic renames in imports', async () => {
        await fetchImportsMessage({ messageBus: messageBusMock, payloads: [{ type: 'FetchImports', importsFromAsset: [{ assetId: 'ASSET#testFinal', keys: ['testNonImportTwo'] }] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

    it('should import multilevel and avoid colliding stub names', async () => {
        await fetchImportsMessage({ messageBus: messageBusMock, payloads: [{ type: 'FetchImports', importsFromAsset: [{ assetId: 'ASSET#testFinal', keys: ['testImportThree'] }] }] })
        expect(JSON.parse(apiClientMock.send.mock.calls[0][0].Data)).toMatchSnapshot()
    })

})