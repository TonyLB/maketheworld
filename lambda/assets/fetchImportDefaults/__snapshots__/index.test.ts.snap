// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchImportsMessage should follow dynamic renames in imports 1`] = `
Object {
  "importsByAsset": Array [
    Object {
      "assetId": "ASSET#testFinal",
      "schemaByKey": Object {
        "testNonImportTwo": "<Room key=(testNonImportTwo)>
    <Exit from=(testNonImportTwo) to=(testImportTwo)>test exit</Exit>
</Room>",
      },
      "stubsByKey": Object {
        "testImportTwo": "<Room key=(testImportTwo) />",
      },
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImportsMessage should import multilevel and avoid colliding stub names 1`] = `
Object {
  "importsByAsset": Array [
    Object {
      "assetId": "ASSET#testFinal",
      "schemaByKey": Object {
        "testImportThree": "<Room key=(testImportThree) />",
      },
      "stubsByKey": Object {
        "testImportAssetThree.stub": "<Room key=(testImportAssetThree.stub)><Name>AssetThreeStub</Name></Room>",
        "testImportAssetTwo.stub": "<Room key=(testImportAssetTwo.stub) />",
      },
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImportsMessage should recursive fetch one level of element and stubs when passed import key 1`] = `
Object {
  "importsByAsset": Array [
    Object {
      "assetId": "ASSET#testFinal",
      "schemaByKey": Object {
        "testImportOne": "<Room key=(testImportOne)>
    <Description>Two</Description>
    <Exit from=(testImportOne) to=(testImportStubOne)>test exit one</Exit>
</Room>",
      },
      "stubsByKey": Object {
        "testImportStubOne": "<Room key=(testImportStubOne) />",
      },
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImportsMessage should return element and stubs when passed non-import key 1`] = `
Object {
  "importsByAsset": Array [
    Object {
      "assetId": "ASSET#testFinal",
      "schemaByKey": Object {
        "testNonImport": "<Room key=(testNonImport)>
    <Description>DescriptionOne</Description>
    <Exit from=(testNonImport) to=(testNonImportStub)>test exit</Exit>
</Room>",
      },
      "stubsByKey": Object {
        "testNonImportStub": "<Room key=(testNonImportStub)><Name>StubOne</Name></Room>",
      },
    },
  ],
  "messageType": "FetchImports",
}
`;

exports[`fetchImportsMessage should return empty when passed no keys 1`] = `
Object {
  "importsByAsset": Array [
    Object {
      "assetId": "ASSET#testFinal",
      "schemaByKey": Object {},
      "stubsByKey": Object {},
    },
  ],
  "messageType": "FetchImports",
}
`;
